### DISCUSSION TITLE
Narrowing existing effects

### DISCUSSION THREAD

**Discussion by wolverian** [haskell-effectful/effectful#148] [Q&A]
Is it idiomatic to want to narrow down the possible actions in an effect? In my specific case, I'd like to expose a `FileSystem` effect that only allows `createDirectory` calls. I can write the effect entirely myself, or I can arrange this with re-exports of Effectful.FileSystem, but I have a few questions:

- Just re-exporting [Effectful.FileSystem.createDirectory](https://hackage.haskell.org/package/effectful-2.2.2.0/docs/Effectful-FileSystem.html#v:createDirectory) allows the user to import [Effectful.FileSystem](https://hackage.haskell.org/package/effectful-2.2.2.0/docs/Effectful-FileSystem.html#t:FileSystem) separately and add other actions, e.g. [removeDirectory](https://hackage.haskell.org/package/effectful-2.2.2.0/docs/Effectful-FileSystem.html#v:removeDirectory). Is there a way to make this safer?
- What about reinterpretation and algebraic effects? Take the `FileSystem` effect in the example from [Effectful.Dispatch.Dynamic](https://hackage.haskell.org/package/effectful-core-2.2.2.2/docs/Effectful-Dispatch-Dynamic.html). I might want to expose a read-only version of this effect; how should I do that?

**Comment by ocharles** [comment]
One option you have is to forget about `effectful` at all here and just pass the function in:

```haskell
module M where

onlyCreateDirectory :: (FilePath -> Eff es ()) -> Eff es ()
onlyCreateDirectory createDirectory = ...

module N where

import M qualified
import Effectful.FileSystem

controller :: FileSystem :> es => Eff es ()
controller = M.onlyCreateDirectory createDirectory
```

**Reply by wolverian** [reply]
Thanks! This inverts my desired relationship between M and N: I want N to not have access to Effectful.FileSystem.

**Comment by dark-ether** [comment]
what would be the problem in creating a separate effect and using an existing effect in its handler with reinterpret?

**Reply by wolverian** [reply]
Verbosity and the opportunity to make mistakes, but perhaps it's the best approach.
